python/tk_framework_desktopserver/server_protocol.py

```diff
@@ -129,56 +181,156 @@ def onMessage(self, payload, is_binary):
         if message["protocol_version"] not in self.SUPPORTED_PROTOCOL_VERSIONS:
             message_host.report_error(
-                "Unsupported protocol version: %s " % message["protocol_version"]
+                "Unsupported protocol version: %s." % message["protocol_version"]
             )
             return
 
         self._protocol_version = message["protocol_version"]
 
+        if message["command"]["name"] == "get_ws_server_id":
+            self._handle_get_ws_server_id(message)
+            return
+
+        # Make sure that nothing gets replied to when encryption is required until the server knows
+        # the public server id and we've turned on encryption by initializing the Fernet instance.
+        if self._is_using_encryption() and not self._fernet:
+            logger.error(self.ENCRYPTION_HANDSHAKE_NOT_COMPLETED[1])
+            self.sendClose(*self.ENCRYPTION_HANDSHAKE_NOT_COMPLETED)
+            return
+
         # origin is formatted such as https://xyz.shotgunstudio.com:port_number
         # host is https://xyz.shotgunstudio.com:port_number
         origin_network = self._origin.lower()
         host_network = self.factory.host.lower()
 
         if self._protocol_version == 2:
-
             # Version 2 of the protocol can only answer requests from the site and user the server
             # is authenticated into. Validate this.
 
             # Try to get the user information. If that fails, we need to report the error.
             try:
                 user_id = message["command"]["data"]["user"]["entity"]["id"]
-            except Exception:
-                logger.exception("Unexpected error while trying to retrieve the user id.")
-                self.sendClose(3000, u"No user information was found in this request.")
+            except Exception as e:
+                logger.exception("Unexpected error while trying to retrieve the user id:")
+                self.sendClose(*self.USER_INFO_NOT_FOUND)
                 return
         else:
             user_id = None
 
-        # If the hosts are different or the user ids are different, report an error.
-        if host_network != origin_network or (user_id is not None and user_id != self.factory.user_id):
-            logger.debug("Browser integration request received a different user.")
-            logger.debug("Desktop site: %s", host_network)
-            logger.debug("Desktop user: %s", self.factory.user_id)
-            logger.debug("Origin site: %s", origin_network)
-            logger.debug("Origin user: %s", user_id)
+        if not self._validate_user(host_network, user_id, origin_network):
             self.factory.notifier.different_user_requested.emit(self._origin, user_id)
-            self.sendClose(
-                3001,
-                u"You are not authorized to make browser integration requests. "
-                u"Please re-authenticate in your desktop application."
-            )
+            self.sendClose(*self.UNAUTHORIZED_USER)
```

Comment:
I was going to ask about the structure of all of these pre-baked error tuples, but than I saw this and it makes good sense. Nice!